File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        comb = [""] * (n * 2)
        res = []

        def backtracking(idx, openCount, closeCount):
            if idx == 2 * n:
                res.append(''.join(comb))
                return

            if openCount < n:
                comb[idx] = "("
                backtracking(idx + 1, openCount + 1, closeCount)

            if closeCount < openCount:
                comb[idx] = ")"
                backtracking(idx + 1, openCount, closeCount + 1)

        backtracking(0, 0, 0)

        return res

File: 0022-generate-parentheses/0022-generate-parentheses.swift
class Solution {
    func generateParenthesis(_ n: Int) -> [String] {
        var combination = Array(repeating: "", count: n * 2)
        var res: [String] = []

        func backtracking(_ currIdx: Int, _ openCount: Int, _ closedCount: Int) {
            if currIdx == n * 2 {
                res.append(combination.joined(separator: ""))
                return
            }

            if openCount < n {
                combination[currIdx] = "("
                backtracking(currIdx + 1, openCount + 1, closedCount)
            }

            if closedCount < openCount {
                combination[currIdx] = ")"
                backtracking(currIdx + 1, openCount, closedCount + 1)
            }
        }

        backtracking(0, 0, 0)

        return res
    }
}